src/Services/BuddyService.cs

```diff
@@ -0,0 +1,213 @@
+            relation2.Status = BuddyStatus.Approved;
+
+            ctx.SaveChanges();
+
+            return true;
+        }
+
+        public BuddyActionResult SetBuddyAsBest(Viking viking, Viking vikingToSet, bool best)
+        {
+            Model.Buddy buddyToSet = ctx.Buddies.Where(e => e.OwnerID == viking.Uid)
+                .FirstOrDefault(e => e.BuddyID == vikingToSet.Uid);
+
+            if (buddyToSet == null) return new BuddyActionResult { Result = BuddyActionResultType.Unknown };
+
+            buddyToSet.BestBuddy = best;
+            ctx.SaveChanges();
+
+            return new BuddyActionResult { Result = BuddyActionResultType.Success };
+        }
+
+        public BuddyLocation GetBuddyLocation(Viking viking)
+        {
+            Model.Buddy buddy = ctx.Buddies.FirstOrDefault(e => e.BuddyID == viking.Uid);
+
+            if (viking.CurrentRoom != null) return new BuddyLocation
+            {
+                Server = config.Value.MMOAdress,
+                ServerPort = config.Value.MMOPort,
+                ServerVersion = "S2X", // always use SmartFox 2X Protocol
+                Zone = "JumpStart", // TODO - put this in config
+                Room = viking.CurrentRoom,
+                UserID = buddy.BuddyID.ToString(),
+                MultiplayerID = 1, // TODO - what the fuck is this
+                AppName = "JSMain" // TODO - figure out how to detect this
+            };
+            else return new BuddyLocation();
+        }
+
+        public string GetOrSetBuddyCode(Viking viking, string codeOverride = "")
+        {
+            char[] charList = { 
+                'A', 
+                'B', 
+                'C', 
+                'D', 
+                'E', 
+                'F', 
+                'G', 
+                'H', 
+                'I', 
+                'J', 
+                'K', 
+                'L', 
+                'M', 
+                'N', 
+                'O', 
+                'P', 
+                'Q', 
+                'R', 
+                'S', 
+                'T', 
+                'U', 
+                'V', 
+                'W', 
+                'X', 
+                'Y', 
+                'Z', 
+                '1',
+                '2', 
+                '3', 
+                '4', 
+                '5', 
+                '6', 
+                '7', 
+                '8', 
+                '9', 
+                '0'
+            };
+
+            Random rnd = new Random();
+
+            if (!string.IsNullOrEmpty(codeOverride) && ctx.Vikings.FirstOrDefault(e => e.BuddyCode == codeOverride) == null) viking.BuddyCode = codeOverride;
+
+            if (viking.BuddyCode == null)
+            {
+                string generatedCode = "";
+                for(var i = 0; i < 5; i++)
+                {
+                    generatedCode = generatedCode + charList[rnd.Next(0, charList.Length)];
+                }
+                viking.BuddyCode = generatedCode;
```

Comment:
check for collision here ... also on big server in this approach it is quite high probability of collision ... maybe it's worth considering something better than random 🤔

Reply:
what do you have in mind here?
probably better to just always check for conflicts but if you have a better method, by all means shoot.

Reply:
something like this... but `do/while` loop seems nicer than `goto` for this